File: algorithms/leetcode/ransomNote.go
package main
import "fmt"


func canConstruct(ransomNote, magazine string) bool {
  m := make(map[rune]int)

  for _, letter := range magazine {
    if _, ok := m[letter]; !ok {
      m[letter] = 1 
    } else {
      m[letter] +=1 
    }

  }
  for _, letter := range ransomNote {
    if _, ok := m[letter]; !ok {
      return false 
    }
    m[letter] -=1
    if m[letter] < 0 {
      return false
    }
  }
  return true
}


func main() {
  
  fmt.Println(canConstruct("aa", "aab"))
  fmt.Println(canConstruct("aa", "ab"))

}

File: algorithms/leetcode/ransom_note.py


def canConstructPython(ransomNote, magazine):
    for letter in ransomNote:
        if letter not in magazine:
            return False
        i = magazine.index(letter)
        magazine = magazine[:i:] + magazine[i+1:]

    return True
def canConstruct(ransomNote, magazine):
    m = {}


    for letter in magazine:
        if letter in m:
            m[letter] +=1
        else:
            m[letter] = 1
    for letter in ransomNote:
        if letter not in magazine:
            return False
        m[letter] -=1
        if m[letter] < 0:
            return False
    return True


def canConstructLeet(ransomNote, magazine):
    chars = [0] * 26 
    for c in magazine:
        chars[ord(c) - ord('a')] +=1
    for c in ransomNote:
        chars[ord(c) - ord('a')] -= 1
        if chars[ord(c) - ord('a')] < 0:
            return False
    return True

if __name__ == "__main__":
    print(canConstruct("aa", "ab"))
    print(canConstructPython("aa", "aab"))
    print(canConstructLeet("aa", "aab"))


